api/include/opentelemetry/logs/logger.h

```diff
@@ -0,0 +1,138 @@
+
+  /* Returns the name of the logger */
+  // TODO: decide whether this is useful and/or should be kept, as this is not a method required in
+  // the specification. virtual nostd::string_view getName() = 0;
+
+  /**
+   * Each of the following overloaded log(...) methods
+   * creates a log message with the specific parameters passed.
+   *
+   * @param name the name of the log event.
+   * @param severity the severity level of the log event.
+   * @param message the string message of the log (perhaps support std::fmt or fmt-lib format).
+   * @param record the log record (object type LogRecord) that is logged.
+   * @param attributes the attributes, stored as a 2D list of key/value pairs, that are associated
+   * with this log.
+   * @throws No exceptions under any circumstances.
+   */
+
+  /* The below method is a logging statement that takes in a LogRecord.
+   * A default LogRecord that will be assigned if no parameters are passed to Logger's .log() method
+   * which should at minimum assign the trace_id, span_id, and timestamp
+   */
+  virtual void log(const LogRecord &record) noexcept = 0;
+
+  /** Overloaded methods for unstructured logging **/
+  void log(nostd::string_view message) noexcept
+  {
+    // TODO: move this log method to the SDK? since it needs to call an SDK method
+    // "isEnabled(Severity)"
+
+    // if(isEnabled(Severity::kDefault)){
+    //   return;
+    // }
+
+    // Set severity to the default then call log(Severity, String message) method
+    log(Severity::kDefault, message);
+  }
+
+  void log(Severity severity, nostd::string_view message) noexcept
+  {
+    // TODO: move this log method to the SDK? since it needs to call an SDK method
+    // "isEnabled(Severity)" if(isEnabled(Severity::kDefault)){
+    //   return;
+    // }
+
+    // TODO: set default timestamp later (not in API)
+    log(severity, message, core::SystemTimestamp(std::chrono::system_clock::now()));
+  }
+
+  void log(Severity severity, nostd::string_view message, core::SystemTimestamp time) noexcept
+  {
+    // TODO: move this log method to the SDK? since it needs to call an SDK method
+    // "isEnabled(Severity)" if(isEnabled(Severity::kDefault)){
+    //   return;
+    // }
+
+    // creates a LogRecord object with given parameters, then calls log(LogRecord)
+    LogRecord r;
+    r.severity  = severity;
+    r.body      = message;
+    r.timestamp = time;
+
+    log(r);
+  }
+
+  /** Overloaded methods for structured logging**/
+  // TODO: separate this method into separate methods since it is not useful for user to create
+  // empty logs
+  template <class T,
+            nostd::enable_if_t<common::detail::is_key_value_iterable<T>::value> * = nullptr>
+  void log(Severity severity       = Severity::kDefault,
+           nostd::string_view name = "",
+           const T &attributes     = {}) noexcept
+  {
+    log(severity, name, common::KeyValueIterableView<T>(attributes));
+  }
+
+  void log(Severity severity,
```

Comment:
I swear I saw a different order of arguments in your screenshot during our offline discussion? 😄 

I like this order: `severity, name, attributes`. This is also in alignment with unstructured message order `severity, message`. This is also what you'd typically see in Android OS logging macros, for example. 👍 

You mentioned some other complexity related to too many methods needed. And I'd like to understand where that other concern originated from. If this has been solved since then, please disregard my comment.

Reply:
(Yup we just updated it after the meeting! 😃)

So the other question was how to support optional fields, like: 
 `log(severity, name, attributes)` 
 `log(severity, attributes)` 
 `log(name, attributes)` 
 `log(attributes)`, etc., 

It seems like we would need to write a separate function for each combination of optional parameters. (Since we won't be allowed to set default values of parameters earlier in the list without setting default values for ones after it too). 

In this example, we just have 3 fields, but we're not sure how it would be done to support all 10?